lib/msf/core/exploit/remote/http_client.rb

```diff
@@ -373,27 +374,95 @@ def send_request_raw(opts = {}, timeout = 20, disconnect = false)
-    if cookie_jar.any?
-      opts = { 'cookie' => cookie_jar.to_a.join(' ') }.merge(opts)
+    origin = "http#{ssl ? 's' : ''}://#{peer}"
+
+    if opts.has_key?('cookie') && opts['cookie'].is_a?(HTTP::CookieJar)
+      opts = opts.merge({ 'cookie' => HTTP::Cookie.cookie_value(opts['cookie'].cookies) })
+    elsif !cookie_jar.empty?(origin)
+      opts = opts.merge({ 'cookie' => HTTP::Cookie.cookie_value(cookie_jar.cookies(origin)) })
     end
 
     res = send_request_raw(opts.merge(cgi: true), timeout, disconnect)
-
     return unless res
 
     if opts['keep_cookies'] && res.headers['Set-Cookie'].present?
-      # XXX: CGI::Cookie (get_cookies_parsed) is hella broken
-      cookie_jar.merge(res.get_cookies.split(' '))
+      cookie_jar.parse(res.headers['Set-Cookie'], origin)
     end
 
     res
   end
 
+  # Empties the contents of +cookie_jar+, preventing any cookies
+  # currently stored in the jar from being used in future requests.
+  #
+  # @return [void]
+  def clear_cookies
+    cookie_jar.clear
+  end
+
+  # Merges the contents of the +cj+ param into the +cookie_jar+ class var.
+  # In the case of Cookies with the same name, domain, and path, the +cj+ cookie will be used.
+  #
+  # @param cj [HTTP::CookieJar] CookieJar to be merged into +cookie_jar+.
+  #
+  # @return [HTTP::CookieJar] cookie_jar
+  def merge_cookie_jar!(cj)
+    cj.cookies.each do |c|
+      cookie_jar.add(c)
+    end
+
+    cookie_jar
+  end
+
+  # Merges the contents of  +cj+ & +cookie_jar+ into a new CookieJar.
+  # In the case of Cookies with the same name, domain, and path, the +cj+ cookie will be used.
+  #
+  # @param cj [HTTP::CookieJar] CookieJar to be merged into a new CookieJar with +cookie_jar+.
+  #
+  # @return [HTTP::CookieJar]
+  def merge_cookie_jar(cj)
+    common_cj = HTTP::CookieJar.new
+    cookie_jar.each { |c| common_cj.add(c) }
+
+    cj.cookies.each do |c|
+      common_cj.add(c)
+    end
+
+    common_cj
+  end
+
+  # Makes a new +HTTP::Cookie+ object with sensible default attributes.
+  #
+  # @return [HTTP::Cookie]
+  def make_default_cookie(cookie_name, cookie_value)
+    HTTP::Cookie.new(
+      {
+        :name => cookie_name,
+        :value => cookie_value,
+        :domain => rhost,
+        :created_at => Time.now,
+        :accessed_at => Time.now,
+        :origin => "http#{ssl ? 's' : ''}://#{rhost}",
+        :secure => ssl, # Are you using HTTPS?
+        :path => '/', # Top level of the domain the cookie applies
+        :for_domain => false, # If true, send cookie to any host in the domain.
+        :httponly => true, # Can the flag be accessed by client side scripts?
+        :expires => nil, # At what dateTime (relevant to the client) should the cookie expire
+        :max_age => nil, # How old should the cookie be when it expires? (Alternative to +:expires+)
+        # If +:expires+ && +:max_age+ are nil, cookie will expire when the current session ends
+      }
+    )
+  end
+
```

Comment:
I think we'd be good to move this entirely into the cookie_jar abstraction class directly

I don't think this violates the law of demeter